_posts/2023-07-xx-scala-as-a-junior-dev.md

```diff
@@ -0,0 +1,510 @@
+
+```scala
+val x = 1                     // x: Int = 1
+val list = List(1, 1, 2)      // list: List[Int] = List(1, 1, 2)
+val str = list.mkString(", ") // str: String = 1, 1, 2
+val longThing =               // longThing: Int = 10
+  List(1, 2, 3, 4, 5)
+    .map(0 until _)
+    .filter(_.sum > 8)
+    .flatten
+    .sum
+// same but without type inference
+val longUglyThing: Int =
+  List[Int](1, 2, 3, 4, 5)
+    .map((x: Int) => 0 until x)
+    .filter((r: Range) => r.sum > 8)
+    .flatten
+    .sum
+```
+
+#### 1.2.2. Mix of object-oriented programming and functional programming
+
+Scala has been a pioneer in the mixing of these two paradigms and has been (and still is!) an inspiration for many languages. We benefit from the modularity of objects while having the safety and expressivity of functional programming. Also, unlike some other functional languages, it allows the usage of mutable variables, which can lead to performance improvements.
+
+#### 1.2.3. Multi-platform
+
+Scala can now run on multiple platforms. The main one is the JVM, meaning that it benefits from the Java ecosystem (which is pretty huge) and from a performant long-term runtime environment. And it will get even more performant with the Loom project (lightweight threads).
+
+But it can also run on the browser thanks to the Scala.js project, bringing interoperability with JavaScript. And it can be run as native code (without the JVM) with Scala Native, which can be useful for low-level programming.
+
+#### 1.2.4. Implicits
+
+Implicits are specific values that will be injected by the compiler when they are required (and if they are in available in the current scope). They are very useful to keep a context and not having to write it each time (reduce boilerplate code).
+
+A nice example is with `Future`. A `Future` is a value that will be computed asynchronously in a separate thread. To do so, we need a thread manager as the `Future` only models the value, and we need to pass it each time we create a `Future`.
+
+```scala
+import scala.concurrent.ExecutionContext.global // a thread manager
+
+Future(1 + 1)(global)
+  .map(_ + 1)(global)
+  .flatMap(x => Future(x + 1))(global)
+```
+
+Which is quite cumbersome. But the execution context can be passed implicitly, so we can write:
+
+```scala
+import scala.concurrent.ExecutionContext.Implicits.global
+
+Future(1 + 1)
+  .map(_ + 1)
+  .flatMap(x => Future(x + 1))
+```
+
+Which is much more pleasant to read. But implicits comes with a trade-off: they can be hard to understand at first, and might lead to unexpected behaviors where the wrong implicit is injected. Especially when there are multiple implicits of the same type in the scope (look at the implicit resolution process to avoid this). Still, it is a very powerful feature that helps in making the code more expressive by masking some boilerplates (I know that I'm giving `global` as execution context, so I don't want to see it each time).
```

Comment:
We can add that by using implicits (we can link to the [Givens](https://blog.rockthejvm.com/scala-3-given-using/) article) we can add powerful new functionality, e.g. type classes